zebra-consensus/src/block/check.rs

```diff
@@ -101,14 +102,37 @@ pub fn subsidy_is_valid(block: &Block, network: Network) -> Result<(), BlockErro
     let height = block.coinbase_height().ok_or(SubsidyError::NoCoinbase)?;
     let coinbase = block.transactions.get(0).ok_or(SubsidyError::NoCoinbase)?;
 
+    // Consensus rule: The total value in zatoshi of transparent outputs from a
+    // coinbase transaction, minus vbalanceSapling, minus vbalanceOrchard, MUST NOT
+    // be greater than the value in zatoshi of block subsidy plus the transaction fees
+    // paid by transactions in this block.
+    // https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+    let transparent_value_balance: Amount = subsidy::general::output_amounts(coinbase)
+        .iter()
+        .sum::<Result<Amount<NonNegative>, AmountError>>()
+        .expect("the sum of all outputs will always be positive")
+        .constrain()
+        .expect("positive value always fit in `NegativeAllowed`");
+    let sapling_value_balance = coinbase.sapling_value_balance().sapling_amount();
+    let orchard_value_balance = coinbase.orchard_value_balance().orchard_amount();
+
+    let block_subsidy: Amount = subsidy::general::block_subsidy(height, network)
+        .expect("a valid block subsidy for this height and network")
+        .constrain()
+        .expect("positive value always fit in `NegativeAllowed`");
+
+    // Consensus rule implementation: block fee must be at least zero.
```

Comment:
Possible Bug: This implementation skips the "transaction fees" part of the consensus rule. If we want to skip a check, we need a detailed code comment. (In the code, not just in the PR.)

You can get the transaction fees from the transaction verifier response:
https://github.com/ZcashFoundation/zebra/blob/9b17bbce3d731da24049af138d39f028e4a06b1a/zebra-consensus/src/transaction.rs#L221-L224

I'm also not sure if the equations you used will validate the consensus rule:

> The rule is:
> `valueBalanceTransparent - valueBalanceSapling - valueBalanceOrchard <= blockSubsidy + fees`
> 
> Lets call `valueBalanceTransparent - valueBalanceSapling - valueBalanceOrchard` just `valueBalance`, then we have:
> `valueBalance - blockSubsidy <= fees`

This step creates an extra subtraction, which can create an extra underflow. But that underflow is not part of the rules, so we would need to detect and ignore it:
> `valueBalance - blockSubsidy - fees <= 0`

I think this step breaks the consensus rule:
> `valueBalance - blockSubsidy` should be at least 0(outputs should at least cover block subsidy), so the `fee` should never be negative to hold the inequality.

This assumption is not guaranteed:
> outputs should at least cover block subsidy

Miners don't have to spend the block subsidy or miner fees, they can destroy it instead:
> As in Bitcoin, the remaining value in the transparent transaction value pool of a non-coinbase transaction is available to miners as a fee. The remaining value in the transparent transaction value pool of a coinbase transaction is destroyed.
https://zips.z.cash/protocol/protocol.pdf#transactions

Can you add a test case where miners destroy the block subsidy and miner fees?